### ISSUE TITLE
Strongly encourage setting "license" and "licenses" in package.json

### ISSUE THREAD

**Issue by garthk** [npm/npm#4473]
Specifying [license] in `package.json` makes it a lot easier for users to find and comply with [GNU Affero] licenses. 

Also giving `licenses` (an array of objects with `type` and `url` keys) helps  users whose legal departments request both the license name *and the text*. See this example from [async]:

      "licenses": [{
        "type": "MIT",
        "url": "https://github.com/caolan/async/raw/master/LICENSE"
      }]

If npm \[had] `WARN`-ed about missing `license` and `licenses` like it did missing `description`, `repository`, and `README`, my current task would be a lot easier. 

Solid `license` and `licenses` adoption might also help Node adoption by large companies. If I find any problematic licenses during this mostly manual search — especially buried amongst the dependencies of framework packages like `express`, `connect`, or `restify` — it'll cause the kind of chaos that \[unfairly] entrenches hostility against new platforms. 

Though `licenses` isn't specified in the [npm package.json spec][license], it *is* supported by the NPM site, as you'll see if you check [async's NPM page][async]. So, I guess we've also got a doc-bug. 

[GNU Affero]: http://www.gnu.org/licenses/agpl-3.0.html
[async]: https://npmjs.org/package/async
[asyncpj]: https://github.com/caolan/async/blob/master/package.json
[license]: https://npmjs.org/doc/json.html#license

**Comment by kemitchell** [comment]
I am working on https://github.com/kemitchell/spdx and https://github.com/kemitchell/spdx-correct. Issues, patches, and comments are welcome.

`spdx` aims to support the new "license expression syntax", including`WITH` and `AND` expressions. `spdx-correct` could helpfully suggest corrections when module authors _nearly_ specify valid SPDX identifiers, like 'Apache 2' or just 'GPL'. That way they can make an unambiguous choices in the moment.

For some idea of how "dirty" `.license` data is at the moment, see this dat project: https://github.com/kemitchell/licenses-on-npm

If you'd rather not wait for dat to slurp in all the metadata, here's a gist of fairly recent results: https://gist.github.com/kemitchell/ad3bdc6ca8888599502e

**Comment by othiym23** [comment]
@kemitchell I lack permissions to view either https://github.com/kemitchell/spdx or https://github.com/kemitchell/spdx-correct – are those the correct names / URLS?

**Comment by kemitchell** [comment]
My mistake, @othiym23. The correct URLs are:
- https://github.com/kemitchell/spdx.js
- https://github.com/kemitchell/spdx-correct.js

**Comment by othiym23** [comment]
@kemitchell That looks very promising, and like a natural counterpart to `semver`.

While the npm core team lacks the resources to get to integrating this ourselves (at least in the near to medium term), I think it would be good for the npm ecosystem were npm to support this. If you wanted to experiment with integrating it yourself, the places to start are [`npm/init-package-json`](https://github.com/npm/init-package-json) and [`normalize-package-data`](https://github.com/npm/normalize-package-data). (It would be great if `init-package-json` used `normalize-package-data` to validate the package file it's about to generate, but that is a separate / separable project.) Once those packages are updated to use SPDX, it's fairly straightforward to get that rolled into npm and to update the documentation appropriately.

TL;DR: we don't have the bandwidth to do this ourselves right now, but I'm strongly +1 on something like this being integrated into npm. Thanks for bringing it to my attention!

**Comment by kemitchell** [comment]
Thanks for your message, @othiym23. I especially appreciate the links to `init-package-json` and `normalize-package-data`.

**Comment by othiym23** [comment]
This is closed by the resolution of #8179 – multiple licenses can be included by using an SPDX license expression, `npm init` now prompts for an SPDX-compliant license expression, and `npm install` warns when `"license"` is missing or isn't parseable as an SPDX license expression. See [the release notes for `npm@2.10.0`](https://github.com/npm/npm/releases/tag/v2.10.0) for details.

**Comment by SimenB** [comment]
@othiym23 Any chance of a feature like @andreineculau mentioned, being able to go `npm ls --licenses` or `npm licenses` to get a list out?

**Comment by othiym23** [comment]
@SimenB That sounds like an excellent idea for a package! Like, say, @davglass's [`license-checker`](https://www.npmjs.com/package/license-checker). :D

**Comment by SimenB** [comment]
@othiym23 Good idea. I get stack trace, though...
```
TypeError: Cannot read property 'indexOf' of null
    at /Users/simen/.nvm/versions/node/v0.12.2/lib/node_modules/license-checker/lib/index.js:80:36
    at Array.forEach (native)
    at flatten (/Users/simen/.nvm/versions/node/v0.12.2/lib/node_modules/license-checker/lib/index.js:76:11)
    at /Users/simen/.nvm/versions/node/v0.12.2/lib/node_modules/license-checker/lib/index.js:104:20
    at Array.forEach (native)
    at flatten (/Users/simen/.nvm/versions/node/v0.12.2/lib/node_modules/license-checker/lib/index.js:95:40)
    at /Users/simen/.nvm/versions/node/v0.12.2/lib/node_modules/license-checker/lib/index.js:104:20
    at Array.forEach (native)
    at flatten (/Users/simen/.nvm/versions/node/v0.12.2/lib/node_modules/license-checker/lib/index.js:95:40)
    at /Users/simen/.nvm/versions/node/v0.12.2/lib/node_modules/license-checker/lib/index.js:118:20
```

But I get the intent, this exists in userland. Thanks for the quick response!

**Comment by kemitchell** [comment]
@SimenB, you may be interested in davglass/license-checker#32.

I am working on my own [jslicense/lincesee.js](https://github.com/jslicense/licensee.js). There is also licenserc/licenserc-specification#3. Both early stages. Input welcome!

**Comment by mzgol** [comment]
npm now warns when doing `npm install` on the jquery repo where we use the `licenses` field:
```
$ npm i
npm WARN package.json jquery@3.0.0-pre No license field.
```

I assume that's a bug?

**Comment by SimenB** [comment]
@mzgol You're supposed to use the singular `license`

**Comment by mzgol** [comment]
> @mzgol You're supposed to use the singular license.
> 
> Replace your licenses with the simpler "license": "MIT"

This would work in jQuery scenario but @othiym23 said in https://github.com/npm/npm/issues/4473#issuecomment-32144111:
> The same enterprise logic leads me to ask why licenses is deprecated; I agree that settling on one or the other is a good idea, but it's not that uncommon for larger, more complex projects to be multi-licensed, and it be useful to be able to capture that information in a machine-parseable format.

So it seems it shouldn't warn?

**Comment by SimenB** [comment]
@mzgol From https://github.com/npm/npm/pull/8179
> "license" should be a string SPDX license expression. Thus "GNU Public License" is not good metadata, while "GPL-3.0" is fine. So is "(MIT OR GPL-2.0)" for a "multi-licensed" project.

**Comment by rlidwka** [comment]
I think the current jQuery `license` field:

```js
  "licenses": [
    {
      "type": "MIT",
      "url": "https://github.com/jquery/jquery/blob/master/LICENSE.txt"
    }
  ]
```

... is correct. I don't know why would someone specify an array consisting of one element, but `npm` should **not** warn about this.

**Comment by othiym23** [comment]
@SimenB is correct. As of `npm@2.10.0`, both `licenses` and using an object or array for the value of `license` are deprecated. The SPDX implementation supersedes a lot of the older discussion of licenses with its support for multiple licenses and its relatively simple expression syntax. The docs that @SimenB links to are pretty comprehensive about the details.

**Comment by mzgol** [comment]
Thanks, I've created an issue so that we fix it before 3.0.0: https://github.com/jquery/jquery/issues/2331

**Comment by rlidwka** [comment]
> both licenses and using an object or array for the value of license are deprecated.

@othiym23 ,

There is no explanation of *why* it is deprecated, and I cannot find a relevant discussion in npm tracker about it.

@mzgol ,

Furthermore, new syntax does not allow to specify a custom flavour of a `MIT` license. For example, jQuery appends a disclaimer about licenses of `node_modules` contents to its license, but the license is still `MIT`.

If jQuery changes to `license: "MIT"`, npm homepage would link to opensource.org displaying the wrong license without that notice (and without jquery copyrights by the way). If jQuery changes to `license: "LicenseRef-LICENSE"`, it'll appear as it is not a free software, which is unacceptable.

Imho, it should stay as is.

**Comment by othiym23** [comment]
> There is no explanation of why it is deprecated, and I cannot find a relevant discussion in npm tracker about it.

`licenses` has been deprecated for a while, and the object format has persisted in a few places (mostly @isaacs's modules), but mostly people have been using `license` for a while now. @kemitchell's proposal to switch npm to use SPDX came with working code and a compelling set of rationales: support for a machine-verifiable syntax, support for multiple licensing, and a whole infrastructure for supporting things like custom and proprietary licenses that exists external to npm. It was a relatively simple change that provides a huge benefit to a large class of npm's users. Including it was pretty much a no-brainer.

When we make breaking changes to npm, we will follow a deprecation policy (add warnings for at least one major version cycle or six months, whichever is longer). This isn't a breaking change. All it does is add a warning (which many users have been requesting for a while now) and some documentation around best practices. I'm comfortable with the CLI team reserving the right to periodically make decisions unilaterally when we think it benefits npm's user community, and I think this is a good example of why that approach makes sense.

> If jQuery changes to license: "LicenseRef-LICENSE", it'll appear as it is not a free software, which is unacceptable.

All `LicenseRef-LICENSE` means is that the license doesn't have a short identifier registered with spdx.org. It _can_ be used for proprietary licenses, but that doesn't mean that it _always_ is used for proprietary licenses. We clearly have some additional work to do to make working with custom licenses more integrated into npm, but for now that's the right thing to do with licenses that don't have short identifiers.

**Comment by kemitchell** [comment]
@rlidwka, thanks for your input here. I know this isn't the most interesting subject in the world.

Since @othiym23 and I worked up the patch, I feel I can speak a bit to the rationale for deprecation. We have a standard, SPDX, at least part of which we're in a position to utilize in npm. Many early npm authors settled on the GemSpec-eque `licenses` array early on, and in fact npm's license metadata documentation was modeled on RubyGems'. I'm not sure why or exactly when npm's docs started recommending singular `license` with a string over plural `licenses` with an array, but that approach has now been widely adopted. [Even among the 1,000 most-depended-upon packages](http://npm1k.org/), many of which are older projects with `package.json` files written long ago, more use `license` with a string than `licenses`. In the Ruby world, where I've also proposed SPDX validation in the package manager at rubygems/rubygems#1249, standardizing on license expressions would be a fight up a much steeper hill, so the patch sticks with arrays and validates contained strings.

As for "custom flavors" of MIT, the trouble here is certainty. The MIT license is a specific piece of legal text, and SPDX `MIT` gives legal departments the ability to make decisions about that specific text and reapply them to packages that use it. SPDX does in fact address variants, such as those used by Enlightenment (`MIT-advertising`), feh (`MIT-feh`), and JSON (`JSON`). It also provides [its own guidance about variations of say, the MIT license, that still count as SPDX `MIT`](http://spdx.org/spdx-license-list/matching-guidelines).

As for `LicenseRef-LICENSE`, it doesn't mean that a package is closed-source, but only that it is using some non-standard license. "Open source" is a concept, [variously expressed](http://opensource.org/osd-annotated) and variously implemented by various well-known license templates. I could draft an entirely new license that is both open-source and completely unlike anything with an SPDX identifier. This happened quite a bit before the general mood turned against "license proliferation" and in favor of reusing existing licenses.

Looking at [jQuery's license file](https://github.com/jquery/jquery/blob/master/LICENSE.txt), I'm not sure the notice you mention, below the second `===` divider, is part of the license per se. I'm not sure that the the first three paragraphs above the first `===` are, either. When in doubt, consult foundation counsel! Please feel free to mention my name and e-mail address, which is available on my website.

**Comment by kemitchell** [comment]
> [...] We clearly have some additional work to do to make working with custom licenses more integrated into npm [...]

@othiym23, if there's an org or a working group on that topic, please add me to it. I'd love to help as I can.

**Comment by othiym23** [comment]
> @othiym23, if there's an org or a working group on that topic, please add me to it. I'd love to help as I can.

You're already there. ;) I'll file more issues as I figure out what features need to be added.

**Comment by rlidwka** [comment]
> We have a standard, SPDX, at least part of which we're in a position to utilize in npm.

There is a standard on everything. Usually there are multiple standards conflicting with each other, because it's how this world works (xkcd://standard+proliferation). If there is a standard, it doesn't mean we need to implement it.

This particular standard has a worthy goal though. It is a good idea to have npm check license names, so automated tools can check all the licenses from all the dependencies of a particular packages and avoid trouble with misspellings.

So unless they get really picky about accepting new licenses (like opensource.org guys did), I do think it's a good idea to implement it.

However, if I understand correctly, SPDX allows you to specify multiple licenses:

```xml
<Package rdf:about="">
 <licenseDeclared>
 <ConjunctiveLicenseSet>
 <member rdf:resource="http://spdx.org/licenses/LGPL-2.0" />
 <member rdf:resource="LicenseRef-3" />
 </ConjunctiveLicenseSet>
 </licenseDeclared>
</Package>
```

If this proposal does not, it's not something worth having. Because you are restricting npm users instead of helping them avoid mistypes.

> The MIT license is a specific piece of legal text, and SPDX MIT gives legal departments the ability to make decisions about that specific text and reapply them to packages that use it.

My choice of words was unfortunate, as I was talking about copyrights, not actual variations (which is a real problem in OSS, looking how JSON license sucks).

The trouble is, if I specify `MIT`, npm website will link to [this](http://opensource.org/licenses/MIT) license with this copyright notice:

```
Copyright (c) <year> <copyright holders>
```

I'm sure there are people for whom it is unacceptable to have this copyright notice instead of the real one. Is it worth changing the license to `LicenseRef-LICENSE` only because of this?

PS: this might be considered an issue with npm website, but I don't know how one could solve it.

> As for LicenseRef-LICENSE, it doesn't mean that a package is closed-source, but only that it is using some non-standard license.

That is a very naive approach, for I hope we all know how automated tools that do licence reports work.

  - This means my project uses MIT, and I don't care about copyrights:

    ```js
    license: "MIT"
    ```

  - This means my project uses MIT, but if you're going to consult a lawyer, it's better to link him to an actual text with copyright information instead:

    ```js
    license: {
      type: "MIT",
      url: "http://example.com/LICENSE"
    }
    ```

  - This means my project uses whatever, and you should consult a lawyer to determine what license it is:

    ```js
    license: "LicenseRef-LICENSE"
    ```

Automated tools will report 3rd one as a closed source one. No doubt about that (*). Thus, it should never be used for an opensource software.

(*) First tool to implement SPDX checking would be simple enough, reporting all unparseable licenses as "bad". Then developers will avoid those licenses to avoid their packages being flagged. Thus, only minority of people will use those links. Which would reinforce the idea that all unparseable licenses are better to be reported as "bad". That's how it always works in software engineering.

PS: I think it could be solved allowing multiple licenses:

```js
licenses: [
  "MIT",
  "http://example.com/LICENSE"
]
```

Which would literally mean "this package is using either MIT or that license, pick one". Thus, automated tools will report it as a MIT, and npm website could specify MIT but link to example.com instead, because it has more precise definition.

But until something like that happens, deprecation notice is a mistake that we as a community should not follow.

**Comment by kemitchell** [comment]
On **multi-licensing**: @rlidwka, forgive me if I haven't been clear: SPDX 2.0 includes a stable release of "license expressions" which do support multi-licensing. Here is the example from the npm documentation:

```json
{ "license" : "(ISC OR GPL-3.0)" }
```

This is functionally equivalent to, though better-defined than, the array approach:

```json
{ "licenses" : [{"type": "ISC"}, {"type": "GPL-3.0"}] }
```

And there are kinds of licensing situations that SPDX expressions capture that arrays do not. The [README to spdx.js](https://github.com/kemitchell/spdx.js), which is also its test suite, contains examples used in the spec. You'll see the `WITH` syntax for bolt-on exceptions as well as the `AND` conjunction at work and `()` grouping.

On **standardization**:

> There is a standard on everything. Usually there are multiple standards conflicting with each other, because it's how this world works (xkcd://standard+proliferation). If there is a standard, it doesn't mean we need to implement it.

Amen, brother! But there is a good rationale for standardization. In this case, the goal is machine-readable license metadata that is consistent and unambiguous. npm doesn't have this now, and so much time is wasted tracking down licenses or glancing LICENSE files. Adopting _any_ standard for metadata yields big benefits.

Is there a viable competitor-standard to SPDX that I'm not aware of? Do send a link!

As for the **npm website**, yes, I do think that's better handled as an issue there. There is certainly value in making sure the centralized repository presents the right picture. And we don't want www "enforcing" or acting upon one standard for data, while the CLI propounds another.

**Comment by rlidwka** [comment]
> SPDX 2.0 includes a stable release of "license expressions" which do support multi-licensing. 

Oh that's great then. All that's left to do is write a code for the website to change this:

```js
{ "license" : "(ISC OR LicenseRef-LICENSE)" }
```

To this:

```html
<a href="http://github.com/blahblah/LICENSE">ISC</a>
```

But I'm not sure it could be done reliably in a general case:

```js
{ "license" : "(ISC OR LicenseRef-LICENSE-ISC) OR (MIT OR LicenseRef-LICENSE-MIT)" }
```

```html
<a href="http://github.com/blahblah/LICENSE-ISC">ISC</a> or <a href="http://github.com/blahblah/LICENSE-MIT">MIT</a>
```

I mean good luck teaching computer to guess which url matches which license. The url to the license file is an information that is lost in transition. This is what's wrong with this idea.

> The README to spdx.js, which is also its test suite, contains examples used in the spec. 

I don't see how to make a link to an external website containing a specific license, but maybe I just missed something.

**Comment by SimenB** [comment]
Regarding the website, I made this https://github.com/npm/normalize-license-data/issues/1. I _think_ that's the correct repo.

On how to extract the correct URL... no idea.

**Comment by kemitchell** [comment]
@rlidwka thanks again for good input! I'd like to respond to these parts:

> The url to the license file is an information that is lost in transition. This is what's wrong with this idea.
> ...
> I don't see how to make a link to an external website containing a specific license, but maybe I just missed something.

npm's adoption of SPDX excludes using URLs. From the point of view of legal analysis and avoiding laborious license checks, I think this is actually the best part. HTML is messy stuff, and not great for computers to read for content. Even given a repository of license template texts, which SPDX has, guessing the license found in just a plain text file is non-trivial. Formatting differs as to indentation, line wrapping, typographer's quotes or straight, &c. In cases like jQuery's (and npm's, for that matter) there is extra verbiage in the LICENSE file that confounds straightforward approaches to recognition, like minimum Levenshtein distance. You end up training a neural net or blindly testing tweaks to a rule set against training data. Big, scary proprietary software companies are selling just this kind of thing now.

To make checks truly automatic, we need an unambiguous way to refer to known licenses, and it needs to be the case that if you have license and other metadata about the package, say about authorship, you don't need a URL or a string or any other precomputed rendering of the corresponding legal text. It becomes trivial to generate the relevant license text in whatever format from the metadata alone, and, more to the point, that's no longer necessary. Legal can make an assessment about `AGPL-1.0`, and there's no need to do round-trips from the copy of `AGPL-1.0` the lawyers read to the license ID to a URL containing a `<pre>` tag with plain text wrapped to 80 instead of the lawyer's 72.

As for permitting dirty input and standardizing on `spdx-correct`, I'm afraid that would make the "standard" not-so-standard, by making it language-dependent. Even if we tried to write that package's rules up in plain language and call it the reference implementation, we'd end up worrying about getting and keeping the English and the JavaScript in sync.

**Comment by kemitchell** [comment]
@SimenB  and @rlidwka, if you're interested in the problem of generating HTML links for complex SPDX expressions, I'd encourage you to have a look at the `parse` function of [spdx.js](https://github.com/kemitchell/spdx.js). The [README section on precedence](https://github.com/kemitchell/spdx.js#order-of-precedence-and-parentheses) has some examples.

**Comment by rlidwka** [comment]
> From the point of view of legal analysis and avoiding laborious license checks, I think this is actually the best part. HTML is messy stuff, and not great for computers to read for content. 

So you're still missing the point I wanted to make about specifying license type for computers **and** license text for humans for the **same** license. Which you could do before with license object (and it is used heavily by people) and cannot do now with SPDX expression.

**Comment by kemitchell** [comment]
Forgive me, but I don't think that's the case. An SPDX identifier refers to specific patterns of license text in a way that people can read and recognize. Thus the MIT License is `MIT`, Apache version 2 is `Apache-2.0`, etc.

A URL points to just one way to display that license text. Because an SPDX expression is unambiguous, a computer could generate the license text in any number of formats, say at the command line. My work on [licensor.js](https://github.com/jslicense/licensor.js), [Uber's licensing utility](https://github.com/uber/uber-licence), and other projects take this approach.

On the other hand, if metadata specifies, say, `MIT` for `license` and an author, but points to a URL that is not quite an MIT license, then use of the `MIT` identifier is incorrect and misleading. `package.json` should instead direct attention to the contents of the package in those cases, because legal conclusions about the MIT License may not hold with respect to the variant being used. Only identity between the SPDX identifier and the license text pattern makes increased automation of the legal review process possible.

Many thanks for continued feedback! It's great to meet another member of the npm community who understands the importance of licensing and metadata.

**Comment by scottgonzalez** [comment]
> Looking at jQuery's license file, I'm not sure the notice you mention, below the second === divider, is part of the license per se. I'm not sure that the the first three paragraphs above the first === are, either. When in doubt, consult foundation counsel! Please feel free to mention my name and e-mail address, which is available on my website.

That's correct. The actual license is just plain old MIT, between the two dividers. The license is written that way because I requested that we do so. Since many developers don't think about dependencies when considering licensing, but they're still bound to the licensing terms, we make it explicitly clear that there are other licenses that come into play in the third section. Also, we want to properly list the copyright holders, but it would be unwieldy to list them all directly in the license file, so we have a note about the authors and copyright holders in the first section. We follow this pattern for all jQuery Foundation projects.

Personally, I would want the license displayed as MIT, but with a link to our actual license file. The current object form for `license` makes this very simple to express.

**Comment by kemitchell** [comment]
@scottgonzalez, thanks very much for confirming the scope of the license!

When you mention a link to the license, do you mean on npmjs.com? I would support doing the same kind of introspection on packages that identifies README files for [inclusion in registry metadata](http://registry.npmjs.com/jquery), and linking to LICENSE files within the repository when the repo is hosted on a server with a known HTTP interface, like GitHub. That URL could be used by the web interface.

@othiym23, do you happen to know where README detection takes place? More or less the same logic, with perhaps a broader filename regexp, could probably do the job.

As for giving fair warning of dependency licenses, fortunately for us, the recursive dependency structure of npm makes that pretty straightforward for automated tools. My own experimental audit tools, [licensee](https://github.com/jslicense/licensee.js/), already does recursive checks. Tools built with node in mind should do automatically what the additional language in LICENSE tells people to do.

**Comment by othiym23** [comment]
> @othiym23, do you happen to know where README detection takes place? More or less the same logic, with perhaps a broader filename regexp, could probably do the job.

It happens in a few places – `read-package-json` pulls it in and puts it on the version of the `package.json` data that gets published to the registry, and `fstream-npm` is what ensures that it gets included in the package tarball. I mention the latter because it already has logic to ensure that `LICENSE`, `LICENCE`, and varieties with different capitalization and extensions also get put into the package tarball.

It would be possible to add this to published package metadat on the registry but for the fact that adding large wads of text to CouchDB documents starts making the database sad, which is why the README gets truncated at 50K. If we want to allow people to pull up the text of the license from the web site, that's probably a feature to be requested for [`newww`](https://github.com/npm/newww) or the [`public-api`](https://github.com/npm/public-api), so that it can be decoupled from the CouchDB bits of the registry.

**Comment by scottgonzalez** [comment]
> When you mention a link to the license, do you mean on npmjs.com?

Yes.

> I would support doing the same kind of introspection on packages that identifies README files

That would work for me.

> As for giving fair warning of dependency licenses, fortunately for us, the recursive dependency structure of npm makes that pretty straightforward for automated tools.

Not really. That only covers dependencies that are resolved via npm. Many projects also include dependencies from other sources, such as Bower or direct downloads.

> Tools built with node in mind should do automatically what the additional language in LICENSE tells people to do.

But again, npm is just one source. And npm itself is not just for node.

**Comment by kemitchell** [comment]
Could we include only the file name of the license, without packing its content?

This might produce temporarily broken links in rare cases where:

1. Package `X`, developed on GitHub, uses `LICENSE` for `X@1.0.0`.

2. Package `X` switches to `LICENSE.md` for upcoming `X@2.0.0`.

3. In the interim, `X` is developed on the default GitHub branch.

Without a Git hash to use in the URL for the 1.0.0 commit, npmjs.com might link to `/blob/master/LICENSE`, instead of `/blob/master/LICENSE.md`, and therefore sends visitors to 404.

I glanced the beta GitHub licenses API docs. Didn't see any sign of a file path for the license file, just the name of the detected license.

**Comment by rlidwka** [comment]
> But again, npm is just one source. And npm itself is not just for node.

Oh right, I forgot one point that `package.json` is used by other libraries as well, and putting SPDX identifier there contradicts even CommonJS 1.0 standard npm claims to support:

> licenses - array of licenses under which the package is provided. This property is not legally binding and does not necessarily mean your package is licensed under the terms you define in this property. Each license is a hash with a "type" property specifying the type of license and a url property linking to the actual text. If the license is one of the [official open source licenses](http://www.opensource.org/licenses/alphabetical) the official license name or its abbreviation may be explicated with the "type" property. If an abbreviation is provided (in parentheses), the abbreviation must be used.
>
> ```json
> "licenses": [
>   {
>       "type": "GPLv2",
>       "url": "http://www.example.com/licenses/gpl.html",
>   }
>]
>```

Thus, third party tools implementing CommonJS (like I don't know, Bower?) would have a problem with this. I know the name differs, but considering the history of npm, one would assume the same data type applies to "license". That would be an additional argument for https://github.com/npm/npm/issues/8222.

**Comment by othiym23** [comment]
I think this discussion has gone down the rat hole a little, so I'm locking the thread. @scottgonzalez, thanks for your input on the needs of jQuery – it's very useful in terms of helping me figure out where we need to take this next to handle some of the corner cases. If you or @kemitchell want to continue this discussion in the context of making npm support jQuery better, please open a new issue and we can get that work prioritized on the road map.
